File: Cpp/gtest_version/leaderboard/Leaderboard.tests.cpp
#include "TestData.hpp"
#include <gtest/gtest.h>

#include <map>
#include <string>

using namespace ::testing;

TEST(Leaderboard, SumThePoints)
{
    std::map<std::string, int> results = TestData::sampleLeaderboard1.driverResults();
    ASSERT_NE(results.end(), results.find("Lewis Hamilton"));
    ASSERT_EQ((18 + 18 + 25), results["Lewis Hamilton"]);  
}

TEST(Leaderboard, FindTheWinner)
{
    ASSERT_EQ(0, TestData::sampleLeaderboard1.driverRankings().front().compare("Lewis Hamilton"));  
}

TEST(Leaderboard, KeepAllDriversWhenSamePoints)
{
    // bug, drops drivers with same points
    Race winDriver1("Australian Grand Prix", TestData::CreateDriverList(TestData::driver1, TestData::driver2, TestData::driver3));
    Race winDriver2("Malaysian Grand Prix", TestData::CreateDriverList(TestData::driver2, TestData::driver1, TestData::driver3));
    Leaderboard exEquoLeaderBoard(TestData::CreateRaceList(winDriver1, winDriver2));

    std::list<std::string> rankings = exEquoLeaderBoard.driverRankings();
    std::list<std::string>::const_iterator r = rankings.begin();
    ASSERT_EQ(TestData::driver2.getName(), *r); ++r;
    ASSERT_EQ(TestData::driver1.getName(), *r); ++r;
    ASSERT_EQ(TestData::driver3.getName(), *r); ++r;
    // note: the order of driver1 and driver2 is runtime/platform dependent
}

File: Swift/Tests/LeaderboardTests/LeaderboardTests.swift
import XCTest
import Leaderboard

final class LeaderboardTests: XCTestCase {
    func test_shouldSumThePoints() {
        let results = TestData.shared.sampleLeaderboard1.driverResults()

        XCTAssertTrue(results.keys.contains("Lewis Hamilton"))
        XCTAssertEqual(18 + 18 + 25, results["Lewis Hamilton"])
    }

    func test_shouldFindTheWinner() {
        XCTAssertEqual(
            "Lewis Hamilton",
            TestData.shared.sampleLeaderboard1.driverRankings()[0]
        )
    }

    func test_shouldKeepAllDriversWhenSamePoints() {
        let winDriver1 = Race(name: "Australian Grand Prix", drivers: [
            TestData.shared.driver1,
            TestData.shared.driver2,
            TestData.shared.driver3
        ])
        let winDriver2 = Race(name: "Malaysian Grand Prix", drivers: [
            TestData.shared.driver2,
            TestData.shared.driver1,
            TestData.shared.driver3
        ])
        let exEquoLeaderboard = Leaderboard(races: [winDriver1, winDriver2])

        let rankings = exEquoLeaderboard.driverRankings()

        XCTAssertEqual(
            [
                TestData.shared.driver2.name,
                TestData.shared.driver1.name,
                TestData.shared.driver3.name
            ],
            rankings
        )
    }
}
